### PR TITLE
[full-ci] Fix #6658: split SideBar into separate view and business logic components

### LINKED ISSUE #6658: Extract right sidebar from files app into web-pkg and make it reusable for other apps
## Status quo
The files app has a right sidebar which shows panels that represent data or actions referencing the current selection model of the files list. At the moment it follows a pattern of panels being defined with a component reference and a getter whether or not the respective panel is enabled for the current state (selection model, current route).

Screenshot 2022-03-24 at 12 59 11

## Motivation / description
We want to reuse the pattern of having a right sidebar with the following layout / user flow:
- header with back-navigation to parent panel, title of the active panel, button to close the sidebar
- if in root:
  - display active root panel
  - display nav items to available subpanels
- if in subpanel:
  - display active panel
  - back to parent button in header is available

Only the contents of the respective panels are specific to an app. Next step for this is to have the sidebar container and panel handling generic, while leaving it to the app to provide content panels.

## Additional requirements
On top of extracting this logic from the files app, it would be good to be able to display multiple active panels below each other at the same time. For the example of the files app, we have the `FileInfo.vue` component which is shown in most root panels and in all possible sub panels. It makes sense to treat that as just another panel.

## Examples
- user management app is supposed to show details of the selected user as root panel and a) editing the profile information, b) assigning group memberships as sub panels
- mediaviewer could have a right sidebar with mimetype specific root panels showing e.g. EXIF data for the currently displayed image.

### PR DESCRIPTION
**Pull request by dschmidt**
## Related Issue

- Fixes #6658 

## Motivation and Context
Make SideBar reusable.

## How Has This Been Tested?

- test environment:
- test case 1:
- test case 2:
- ...

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:

- [ ] Split test as well, so far the old test has been adjusted to work with the new separation. We should have separate tests for view and business logic
- [ ] cleanup passed props/interface
- [ ] do not handle state in vuex files namespace

### CODE REVIEW

File: packages/web-pkg/src/components/sidebar/SideBar.vue [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+    isContentDisplayed: {
+      type: Boolean,
+      required: false,
+      default: true
+    },
+    // FIXME: remove
+    isSingleResource: {
+      type: Boolean,
+      required: true,
+    }
```
Comment (dschmidt) [inline packages/web-pkg/src/components/sidebar/SideBar.vue:116]: This does not belong here, we need to come up with a better way...

File: packages/web-pkg/src/components/sidebar/SideBar.vue [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+  beforeDestroy() {
+    visibilityObserver.disconnect()
+    hiddenObserver.disconnect()
+  },
+  methods: {
+    ...mapActions('Files/sidebar', {
+      closeSidebar: 'close',
+      setSidebarPanel: 'setActivePanel',
+      resetSidebarPanel: 'resetActivePanel'
+    }),
```
Comment (dschmidt) [inline packages/web-pkg/src/components/sidebar/SideBar.vue:171]: We should not track the state in the files app vuex namespace.

Should we use the store at all? Personally I'd like to store the state in the url... but I see that not everyone might like that behavior (sidebar stays open after page reload)

File: packages/web-app-files/src/components/SideBar/SideBar.vue [hunk 3]
```diff
@@ -234,71 +134,13 @@ export default {
...
     if (!this.areMultipleSelected) {
       await this.fetchFileInfo()
     }
-    this.$nextTick(this.initVisibilityObserver)
-  },
-
-  beforeDestroy() {
-    visibilityObserver.disconnect()
-    hiddenObserver.disconnect()
+    this.$nextTick(this.$refs.sidebar.initVisibilityObserver())
```
Comment (dschmidt) [inline packages/web-app-files/src/components/SideBar/SideBar.vue:141]: Can we do this better?

We need to wait for `fetchFileInfo()` to finish before the list of panels is known, that's why we can only initialize the observer then... 
If we need to do it like this, should we maybe invoke `initVisibilityObserver()` by default in `created()` of the SideBar component in `web-pkg` and add an option to opt-out and then do it yourself here? I'd not expect I'd have to call this function usually, so making it mandatory by default seems weird ...

File: packages/web-pkg/src/components/sidebar/SideBar.vue [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+  data() {
+    return {
+      focused: undefined,
+      oldPanelName: null,
+      selectedFile: {},
+    }
+  },
+
+  computed: {
+    ...mapState('Files/sidebar', { sidebarActivePanel: 'activePanel' }),
```
Comment (dschmidt) [inline packages/web-pkg/src/components/sidebar/SideBar.vue:128]: Bad... Todo

**Comment by ownclouders** [comment]
Results for <strong>oCISSharingAndUpload</strong> https://drone.owncloud.com/owncloud/web/24154/66/1
<details><summary>:boom: The acceptance tests failed on retry. Please find the screenshots inside ...</summary>

<p>


### webUIUpload-upload_feature-L94.png
  


</p></details>
:boom: The acceptance tests pipeline failed. The build has been cancelled.

**Review by pascalwengerter — COMMENTED** [review]
Code LGTM, some questions/comments and too tired to approve this now (even though CI seems to look fine so far 😎 )

File: packages/web-app-files/src/App.vue [hunk 5]
```diff
@@ -9,30 +9,37 @@
...
-export default {
+export default defineComponent({
   components: {
     SideBar
   },
   mixins: [Mixins],
   computed: {
-    ...mapGetters('Files', ['dropzone']),
-    ...mapState('Files/sidebar', { sidebarClosed: 'closed' }),
+    ...mapGetters('Files', ['dropzone', 'inProgress']),
```
[the remark this answers is not in the document]
Comment (dschmidt) [reply packages/web-app-files/src/App.vue:38]: Bahaha oh noes. The diff really confused me...

File: packages/web-app-files/src/helpers/share/share.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
+import type { User } from '../user'
+import type { ShareRole } from './role'
+import type { SharePermission } from './permission'
+
+export interface Share {
+  shareType: number
```
[the remark this answers is not in the document]
Comment (dschmidt) [reply packages/web-app-files/src/helpers/share/share.ts:6]: Possibly. All these resource related interfaces need cleanup or even better: a concept. For now they just replicate what we already had and I did not bother to sort props...

File: packages/web-pkg/src/components/sidebar/types.ts [hunk 7]
```diff
@@ -0,0 +1,13 @@
+import { Component } from 'vue'
+
+export type IconFillType = 'fill' | 'line' | 'none'
```
[the remark this answers is not in the document]
Comment (dschmidt) [reply packages/web-pkg/src/components/sidebar/types.ts:3]: This is needed for oc-icon from ODS, so I'd argue it really should go to ODS...

File: packages/web-app-files/src/helpers/share/share.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (kulmann) [reply packages/web-app-files/src/helpers/share/share.ts:6]: legit 👍

File: packages/web-app-files/src/helpers/resource/resource.ts [hunk 8]
```diff
@@ -6,10 +8,47 @@ export interface Resource {
...
   path: string
-  webDavPath: string
+  webDavPath?: string
   downloadURL?: string
   type?: string
   status?: number
+  spaceRoles?: any[]
+  mimeType?: string
+  isFolder?: boolean
+  sdate?: string
```
[the remark this answers is not in the document]
Comment (kulmann) [reply packages/web-app-files/src/helpers/resource/resource.ts:18]: ah nvm, saw your other comment about just replicating what we currently have. fine by me.

**Review by kulmann — APPROVED** [review]
nice 🥳

File: packages/web-app-files/src/helpers/resource/resource.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (dschmidt) [reply packages/web-app-files/src/helpers/resource/resource.ts:18]: Yep, agreed. 

It's pretty shitty currently, but imho it helps identifying, where we are pretty messy with types in the first place. So yeah, I would very much like to see this improved, but we need to start somewhere :)

**Comment by dschmidt** [comment]
Oops, I wanted to squash ...

**Pull request merged**